include/envoy/http/header_map.h

```diff
@@ -651,30 +723,91 @@ class RequestOrResponseHeaderMap : public virtual HeaderMap {
+  virtual void appendInline(Handle handle, absl::string_view data,
+                            absl::string_view delimiter) PURE;
+  virtual void setReferenceInline(Handle, absl::string_view value) PURE;
+  virtual void setInline(Handle, absl::string_view value) PURE;
+  virtual void setInline(Handle, uint64_t value) PURE;
+  virtual size_t removeInline(Handle handle) PURE;
+};
 using RequestTrailerMapPtr = std::unique_ptr<RequestTrailerMap>;
 
 // Base class for both response headers and trailers.
-class ResponseHeaderOrTrailerMap : public virtual HeaderMap {
+class ResponseHeaderOrTrailerMap {
 public:
+  virtual ~ResponseHeaderOrTrailerMap() = default;
+
   INLINE_RESP_HEADERS_TRAILERS(DEFINE_INLINE_HEADER)
 };
 
 // Response headers.
 class ResponseHeaderMap : public RequestOrResponseHeaderMap, public ResponseHeaderOrTrailerMap {
 public:
   INLINE_RESP_HEADERS(DEFINE_INLINE_HEADER)
+
+  // Custom header manipulation via strongly typed handle.
+  // TODO(mattklein123): Possibly define via template mix-in.
+  using Handle = CustomInlineHeaderRegistry<ResponseHeaderMap>::Handle;
+  virtual const HeaderEntry* getInline(Handle handle) const PURE;
+  virtual void appendInline(Handle handle, absl::string_view data,
+                            absl::string_view delimiter) PURE;
+  virtual void setReferenceInline(Handle, absl::string_view value) PURE;
+  virtual void setInline(Handle, absl::string_view value) PURE;
+  virtual void setInline(Handle, uint64_t value) PURE;
+  virtual size_t removeInline(Handle handle) PURE;
 };
 using ResponseHeaderMapPtr = std::unique_ptr<ResponseHeaderMap>;
 
 // Response trailers.
-class ResponseTrailerMap : public virtual HeaderMap, public ResponseHeaderOrTrailerMap {};
+class ResponseTrailerMap : public ResponseHeaderOrTrailerMap, public HeaderMap {
+public:
+  // Custom header manipulation via strongly typed handle.
+  // TODO(mattklein123): Possibly define via template mix-in.
+  using Handle = CustomInlineHeaderRegistry<ResponseTrailerMap>::Handle;
+  virtual const HeaderEntry* getInline(Handle handle) const PURE;
+  virtual void appendInline(Handle handle, absl::string_view data,
+                            absl::string_view delimiter) PURE;
+  virtual void setReferenceInline(Handle handle, absl::string_view value) PURE;
+  virtual void setInline(Handle handle, absl::string_view value) PURE;
+  virtual void setInline(Handle handle, uint64_t value) PURE;
+  virtual size_t removeInline(Handle handle) PURE;
+};
 using ResponseTrailerMapPtr = std::unique_ptr<ResponseTrailerMap>;
 
+class CustomInlineHeaderUtility {
+public:
+  /**
+   * Finalize all concrete header map registrations.
+   */
+  static void finalize() {
+    CustomInlineHeaderRegistry<RequestHeaderMap>::finalize();
+    CustomInlineHeaderRegistry<RequestTrailerMap>::finalize();
+    CustomInlineHeaderRegistry<ResponseHeaderMap>::finalize();
+    CustomInlineHeaderRegistry<ResponseTrailerMap>::finalize();
```

Comment:
Ah I see now there are exactly 4 registries and they are all initialized here. WDYT of having this wrapper-class capture all the static state and having this one capture all the data that we are constructing-on-first-use in the template class?

Reply:
> WDYT of having this wrapper-class capture all the static state and having this one capture all the data that we are constructing-on-first-use in the template class?

Yes I think I can likely collapse all of the public access into a single class. It's possible I can even have the bifurcation be using an enum vs. the interface, which would make it even more clear, but I need to see about that, as there are other follow on template implications. If we decide to pursue this direction I can probably simplify this.